Add CI support for SQLAlchemy 2.0

Dev-iL opened the pull request:
closes: #48953, #52663
related: #28723, #50960 (prerequisite that's included in this PR)

- Add the `check_upgrade_sqlalchemy` container entrypoint for testing SQLA v2 on CI
- Fix some existing shellcheck violations reported by shellcheck-py.
- Robustify multiple unit tests and amend some business logic.

<!--
 Licensed to the Apache Software Foundation (ASF) under one
 or more contributor license agreements.  See the NOTICE file
 distributed with this work for additional information
 regarding copyright ownership.  The ASF licenses this file
 to you under the Apache License, Version 2.0 (the
 "License"); you may not use this file except in compliance
 with the License.  You may obtain a copy of the License at

   http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing,
 software distributed under the License is distributed on an
 "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 KIND, either express or implied.  See the License for the
 specific language governing permissions and limitations
 under the License.
 -->

<!--
Thank you for contributing! Please make sure that your code changes
are covered with tests. And in case of new features or big changes
remember to adjust the documentation.

Feel free to ping committers for the review!

In case of an existing issue, reference it using one of the following:

closes: #ISSUE
related: #ISSUE

How to write a good git commit message:
http://chris.beams.io/posts/git-commit/
-->



<!-- Please keep an empty line above the dashes. -->
---
**^ Add meaningful description above**
Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/contributing-docs/05_pull_requests.rst#pull-request-guidelines)** for more information.
In case of fundamental code changes, an Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvement+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in a newsfragment file, named `{pr_number}.significant.rst` or `{issue_number}.significant.rst`, in [airflow-core/newsfragments](https://github.com/apache/airflow/tree/main/airflow-core/newsfragments).

Dev-iL committed:
CI: Add SQLA2 core & providers workflows

Dev-iL commented on scripts/docker/entrypoint_ci.sh:
```diff
@@ -271,6 +271,22 @@ function check_boto_upgrade() {
+    # shellcheck disable=SC2086
+    ${PACKAGING_TOOL_CMD} uninstall ${EXTRA_UNINSTALL_FLAGS} flask-appbuilder || true
+    # shellcheck disable=SC2086
+    ${PACKAGING_TOOL_CMD} install ${EXTRA_INSTALL_FLAGS} --upgrade "sqlalchemy[asyncio]<2.1" databricks-sqlalchemy
```
Resolving because following an existing convention and refactoring it everywhere is out of scope.

Dev-iL commented on scripts/docker/entrypoint_ci.sh:
```diff
@@ -271,6 +271,22 @@ function check_boto_upgrade() {
 }
 
+# Upgrade sqlalchemy to the latest version to run tests with it
+function check_upgrade_sqlalchemy() {
```
Resolving because following an existing convention and refactoring it everywhere is out of scope.

potiuk replied:
Actually - it's not duplicated. pre-commit is inlining the scripts into Dockerfile.ci automatically :)

Dev-iL commented on scripts/docker/entrypoint_ci.sh:
```diff
@@ -197,14 +197,14 @@ function handle_mount_sources() {
 function determine_airflow_to_use() {
     USE_AIRFLOW_VERSION="${USE_AIRFLOW_VERSION:=""}"
-    if [[ ${USE_AIRFLOW_VERSION} == "" && ${USE_DISTRIBUTIONS_FROM_DIST=} != "true" ]]; then
+    if [[ "${USE_AIRFLOW_VERSION}" == "" && ${USE_DISTRIBUTIONS_FROM_DIST=} != "true" ]]; then
```
Done

Dev-iL commented on scripts/docker/entrypoint_ci.sh:
```diff
@@ -271,6 +271,22 @@ function check_boto_upgrade() {
 
+# Upgrade sqlalchemy to the latest version to run tests with it
+function check_upgrade_sqlalchemy() {
+    if [[ ${UPGRADE_SQLALCHEMY=} != "true" ]]; then
```
Ok

Dev-iL commented on Dockerfile.ci:
```diff
@@ -1063,6 +1063,21 @@ function check_boto_upgrade() {
 }
 
+function check_upgrade_sqlalchemy() {
+    if [[ ${UPGRADE_SQLALCHEMY=} != "true" ]]; then
```
ok

uranusjr commented on airflow-core/tests/unit/api_fastapi/conftest.py:
```diff
@@ -117,6 +117,7 @@ def create_test_client(apps="all"):
 def configure_git_connection_for_dag_bundle(session):
     # Git connection is required for the bundles to have a url.
+    clear_db_connections(False)
     connection = Connection(
```
```suggestion
    clear_db_connections(False)
    # Git connection is required for the bundles to have a url.
    connection = Connection(
```

Nit (keeping the comment directly above the relevant code)

uranusjr commented on devel-common/src/tests_common/pytest_plugin.py:
```diff
@@ -210,6 +210,17 @@ def _calculate_provider_deps_hash():
+        from tests_common.test_utils.mock_plugins import mock_plugin_manager
+
+        with mock_plugin_manager() as _fixture:
+            yield _fixture
```
```suggestion
MOCK_PLUGINS_MANAGER = os.environ.get("_AIRFLOW_SKIP_DB_TESTS", "false") == "true" and importlib.util.find_spec("airflow")


@pytest.fixture(scope="session", autouse=MOCK_PLUGINS_MANAGER)
def mock_plugins_manager_for_all_non_db_tests():
    from tests_common.test_utils.mock_plugins import mock_plugin_manager

    with mock_plugin_manager() as _fixture:
        yield _fixture
```

Something like this

Dev-iL committed:
Remove the upper limit of SQLA

+ Fix some SQLA2 incompatibilities
+ Skip multiple tests that rely on FAB

potiuk commented on .github/workflows/special-tests.yml:
```diff
@@ -135,6 +135,54 @@ jobs:
+      test-scope: "DB"
+      test-group: "core"
+      backend: "postgres"
+      python-versions: "['${{ inputs.default-python-version }}']"
```
You need to change and below to "['3.13']"

potiuk replied:
Hardcoding in this case is fine - but we need to add the comment why we are hardcoding it and when we should change it back (when we migrate to fab 5)

potiuk commented on .github/workflows/special-tests.yml:
```diff
@@ -135,6 +135,54 @@ jobs:
+      debug-resources: ${{ inputs.debug-resources }}
+      skip-providers-tests: ${{ inputs.skip-providers-tests }}
+      use-uv: ${{ inputs.use-uv }}
+
```
And you ned to add "if" condition - to run only when `3.13` is in `input.python_versions`

potiuk replied:
likely something like `if: contains(fromJson(input.python-versions), '3.13')`

potiuk replied:
https://docs.github.com/en/actions/reference/evaluate-expressions-in-workflows-and-actions#contains => input.python-versions is a stringified json array of python versions, fromJson should convert it back to array and contains should check if `3.13` is one of the elements

potiuk commented on .github/workflows/special-tests.yml:
```diff
@@ -135,6 +135,60 @@ jobs:
+      debug-resources: ${{ inputs.debug-resources }}
+      skip-providers-tests: ${{ inputs.skip-providers-tests }}
+      use-uv: ${{ inputs.use-uv }}
+    if: ${{ contains(fromJSON(input.python-versions), '3.13') }}  # Remove this line after upgrading FAB to v5
```
If does not need `${{ }}`

Dev-iL replied:
Doesn't work either way

Dev-iL committed:
Add a 30s start period to all containers to reduce test flakiness

Dev-iL committed:
Bump mysql-connector-python to 9.1.0

Might help with the MySQL 8.4 + Py3.13 combination

Dev-iL committed:
Set MySQL's `isolation_level` to `AUTOCOMMIT` in SQLA2 during DDL CREATE

Dev-iL committed:
Update the `db.py` tests

- Make some tests more specific so as not trip on irrelevant things.
- Add a couple of fixtures to ensure a predictable engine state.
- Replace uses of unittest.mock with the mocker fixture.
- Add tests for `AutocommitEngineForMySQL`.
